pydra/engine/specs.py

```diff
@@ -449,6 +450,35 @@ def _field_defaultvalue(self, fld, output_dir):
 
     def _field_metadata(self, fld, inputs, output_dir):
         """Collect output file if metadata specified."""
+        if "requires" in fld.metadata:
+            field_required = fld.metadata["requires"]
+        # if the output field doesn't have requires field, we use fields from output_file_template
+        elif "output_file_template" in fld.metadata:
```

Comment:
shouldn't this not be an elif and make sure that `requires` is a superset of the keys in `output_file_template` when both are provided? 

i.e. the intersection of keys from `output_file_template` and `requires` is equal to the keys from `output_file_template`.

Reply:
yes, this makes sense, will change

Reply:
but `output_file_template` can have names of outputs... I could select the ones that are in the `input_spec`, or just expect users to add all required inputs to the `requires` list

Reply:
consider whatever is necessary to validate and/or provide the user with a clear error message.

Reply:
but this is related to my other question I asked yesterday. Let me repeat and rephrase:

If a field from `requires` list (assuming for now that it also takes what is needed from the template) is not present:
a) we raise an exception (with clear message)
b) we set `NOTHING` for this specific output field

I was assuming that the answer is `b`, but from you questions I'm guessing that you're thinking about `a`... 
I really think that `b` makes more sense, because:
- we can always set an input field as a `mandatory` if you want to have exceptions when it's not set
- as we were discussing, many interfaces from Nipype has multiple output that is created only for a specific combination for the input, and I was hoping that `requires` can be used to solve this issue, and not to raise errors.

Reply:
yes, outputs should be set to ATTR.nothing when requirements aren't satisfied. 

my hope was to provide appropriate error messages when output_file_templates are used with/without requires:
1. without requires: is the output generated if some input is missing? i would say without requires an output is always created based on output_file_template. and should provide appropriate error message to user when an input to the template is missing.
2. with requires: the output is only generated when requirements are satisfied.`output_file_template` can still error when used and should again generate appropriate error message.

Reply:
doesn't this contradict what you wrote yesterday - " the intersection of keys from output_file_template and requires is equal to the keys from output_file_template"?

I've made changes today and added all fields from ` output_file_template` to `requires`..

Reply:
but as for the idea, sure, we can do it

Reply:
i think it's a bit different:

1. intersection, when `requires` exists, and is satisfied, should also have the `output_file_template` fields. i.e. if the requirements are satisfied then the `output_file_template` fields are also required. 

2. if it's just `output_file_template` without any `requires`, then i agree with your interpretation that the output should only be created if all `args` are available. to enforce generation when requirements are satisfied, the task creator should always add the `requires` metadata to output fields.

Reply:
the second point is very unclear to me:
- what are you calling `args`
- what you mean in " to enforce generation when requirements are satisfied, the task creator should always add the requires metadata to output fields."?  Do you meant that all fields from `output_file_template` should be added to `requires` (that is not provided, so can be treated as an empty list). Isn't this the same as in point 1?

What about the following strategy?
- If you want to have an input that has to be provided - use `mandatory=True` for the input field (if the input not provided the exception will be raised)
- If you want to have a specific combination of the input you can use `requires` and `xor` for the input fields (if the provided input doesn't match `requires/xor` an exception is raised)
- if the specific input, say `inp_notmand` is not mandatory, but without the input field the output field is not created, then the output field should have `requires=["inp_notmand"]`. If `inp_notmand` is not provided the output is not created (no exception). No `requires` is treated as `requires=[]`.
- the template should only include fields from the `requires` list and fields that are always provided (ideally they should have `mandatory=True`). If all fields that have `mandatory=True` are provide, pydra checks `requires`. If all fields from the list are set, then it tries to create the name. If there is still something missing, then we raise the error.

Reply:
i'm fine with the strategy. i don't think anything changes except for how `requires` is assigned for something that only has an `output_file_template` and no `requires`.  if we are to say that `output_file_template` should always have requires, then it goes back to simply adding `output_file_template` format args to `requires`.

Reply:
maybe I will keep for now as it is (i.e. no errors are raised, just output is not created). I really don't like the idea that fields from `output_file_template` are treated differently when `requires` is present. 
My strategy from yesterday has a problem, that I had to add to many `requires` to the tests, and I don't like it anymore...
Let me know if you have strong objections, I never fully understood why it is important to you that `output_file_template` can lead to an error